Support auto provision of device to application

Steve-Mcl opened the pull request:
closes #3515 

## Description

Support auto-assigning an auto provisioned device directly to an application.

![chrome_6tmsoO5zoX](https://github.com/user-attachments/assets/cc4028a3-3eea-4a91-9656-0ac3bb4aecbc)

### Update Tests
```
  AccessToken controller
    Device Provisioning Tokens
      ✔ creates a provisioning token for a team
      ✔ creates a provisioning token for a teams project
      ✔ edits a provisioning token to set the project
      ✔ edits a provisioning token to remove the project
  Team Devices API
    Provisioning Tokens
      ✔ Get list of provisioning tokens for the team
```
### Add new tests
```
  AccessToken controller
    Device Provisioning Tokens
      ✔ creates a provisioning token for a teams application
      ✔ edits a provisioning token to set the application
      ✔ edits a provisioning token to remove the application
  Device API
    Create device
      ✔ creates a device in the team and assigns it to an application using a provisioning token (73ms)
  Team Devices API
    Provisioning Tokens
      ✔ Cannot generate a provisioning token for application in another team
      ✔ Edit a provisioning token to assign an application
      ✔ Edit a provisioning token to unassign an application
      ✔ Cannot edit a provisioning token to assign an application that does not exist
      ✔ Cannot edit a provisioning token to assign an application that is not in the team
```

## Related Issue(s)

Owner:  #3515 
Parent: #5146

## Checklist

<!-- https://flowfuse.com/handbook/development/#defining-done -->

 - [x] I have read the [contribution guidelines](https://github.com/FlowFuse/flowfuse/blob/main/CONTRIBUTING.md)
 - [x] Suitable unit/system level tests have been added and they pass <!-- If not adding test coverage, please clarify why not? -->
 - [ ] Documentation has been updated
    - [ ] Upgrade instructions
    - [ ] Configuration details
    - [ ] Concepts
 - [ ] Changes `flowforge.yml`?
    - [ ] Issue/PR raised on `FlowFuse/helm` to update ConfigMap Template
    - [ ] Issue/PR raised on `FlowFuse/CloudProject` to update values for Staging/Production

## Labels

 - [ ] Includes a DB migration? -> add the `area:migration` label

Steve-Mcl commented:
Acceptance criteria:

- [ ] Generate 3 tokens
   - [ ] No auto assignment
   - [ ] auto assign to instance
   - [ ] auto assign to application
- [ ] Start device agent with a provisioning `device.yml` in the `/opt/flowfuse-device` directory that auto assigns to an instance
- [ ] Start device agent with a provisioning `device.yml` in the `/opt/flowfuse-device` directory that auto assigns to an application
- [ ] Edit "No auto assignment" to point to an Application and test by provisioning a device

Steve-Mcl commented:
NOTE: I would like to point out there are existing mismatches between the docs and the UI & legacy references to "Devices" littered through out both the docs and the UI

e.g. 
![image](https://github.com/user-attachments/assets/83a41389-fed1-4d7b-b18c-39869dd16faf)

![image](https://github.com/user-attachments/assets/f91797fd-43c9-464d-b7dd-7bfb7f3f2c29)


These will be taken care of separately in a branch based off this branch in sub task #5395 where I will square up both the docs and the legacy UI wording

hardillb commented on forge/db/controllers/AccessToken.js:
```diff
@@ -174,15 +174,19 @@ module.exports = {
         const generatedToken = generateToken(32, 'ffadp')
         const scope = ['device:provision', `name:${name}`]
-        const projectId = (project && typeof project === 'object') ? project.id : project
+        const projectId = autoAssignType === 'instance' ? (autoAssignId && typeof autoAssignId === 'object') ? autoAssignId.id : autoAssignId : null
+        const applicationId = autoAssignType === 'application' ? (autoAssignId && typeof autoAssignId === 'object') ? autoAssignId.hashid : autoAssignId : null
```
It might just be me, but these double shortcut if statements are REALLY hard to read/understand

hardillb replied:
I assume the are evaluated in to out?
e.g.

```
const projectId = autoAssignType === 'instance' ?
       (autoAssignId && typeof autoAssignId === 'object') ? autoAssignId.id : autoAssignId 
: null
```

Steve-Mcl replied:
Fair comment. I have altered this to be more verbose and clear in [code readability: remove double ternary logic](https://github.com/FlowFuse/flowfuse/pull/5394/commits/8d87abe8ba5bfdc9da4b31a68deecb497025f471)
[8d87abe](https://github.com/FlowFuse/flowfuse/pull/5394/commits/8d87abe8ba5bfdc9da4b31a68deecb497025f471)

hardillb commented on forge/db/controllers/AccessToken.js:
```diff
@@ -202,16 +206,21 @@ module.exports = {
         let scope = [...(token.scope || [])]
-        const projectId = (project && typeof project === 'object') ? project.id : project
-        // remove project scope & add updated project scope (if set)
+        const instanceId = autoAssignType === 'instance' ? (autoAssignId && typeof autoAssignId === 'object') ? autoAssignId.id : autoAssignId : null
+        const applicationId = autoAssignType === 'application' ? (autoAssignId && typeof autoAssignId === 'object') ? autoAssignId.hashid : autoAssignId : null
```
Same as above

Steve-Mcl replied:
Fair comment. I have altered this to be more verbose and clear in [code readability: remove double ternary logic](https://github.com/FlowFuse/flowfuse/pull/5394/commits/8d87abe8ba5bfdc9da4b31a68deecb497025f471)
[8d87abe](https://github.com/FlowFuse/flowfuse/pull/5394/commits/8d87abe8ba5bfdc9da4b31a68deecb497025f471)

hardillb commented on forge/routes/api/teamDevices.js:
```diff
@@ -170,6 +171,9 @@ module.exports = async function (app) {
         const team = request.team
         const instanceId = request.body.instance
+        const applicationId = request.body.application
+        const autoAssignType = instanceId ? 'instance' : (applicationId ? 'application' : null)
```
This one is easier to understand with the brackets

hardillb commented on forge/routes/api/teamDevices.js:
```diff
@@ -202,7 +213,7 @@ module.exports = async function (app) {
             }
             const resp = { code: err.code || 'unexpected_error', error: responseMessage }
-            await app.auditLog.Team.team.device.provisioning.created(request.session.User, resp, '', tokenName, team, instanceId)
+            await app.auditLog.Team.team.device.provisioning.created(request.session.User, resp, '', tokenName, team, instanceId) // TODO: add applicationId
```
Did this TODO get done?

Steve-Mcl replied:
Handled here: [include application parameter in provisioning created log](https://github.com/FlowFuse/flowfuse/pull/5394/commits/42c3b9306971f8cd81ad78202219fcee076fda94) and here: [Update audit logging for provisioning tokens to include application d…](https://github.com/FlowFuse/flowfuse/pull/5394/commits/3b889fb55c1f8cad75ce1e2ecefd7a7fe3551e76)

hardillb reviewed (commented):
I've tested and a it all appears to work.

I've not approved just yet for 2 reasons

1. The comments about readability of the nested shortcut if statements
2. The `//TODO` that appears to have not been done.

We can ship with 1 and I'll approve if 2 is not something you meant to go back to?

Steve-Mcl commented:
@hardillb - I addressed your concerns (hopefully)

I did a quick retest on pre-staging (seems to work as before)

Hopefully tests are still passing
